packages/teleterm/src/preload.ts

```diff
@@ -38,6 +48,38 @@ async function getElectronGlobals(): Promise<ElectronGlobals> {
     mainProcessClient,
     tshClient,
     ptyServiceClient,
-    loggerService,
+  };
+}
+
+/**
+ * For TCP transport, createGrpcCredentials generates the renderer key pair and reads the public key
+ * for tshd and the shared process from disk. This lets us set up gRPC clients in the renderer
+ * process that connect to the gRPC servers of tshd and the shared process.
+ */
+async function createGrpcCredentials(
+  runtimeSettings: RuntimeSettings
+): Promise<{
+  // Credentials for talking to the tshd process.
+  tshd: ChannelCredentials;
+  // Credentials for talking to the shared process.
+  shared: ChannelCredentials;
+}> {
+  if (!shouldEncryptConnection(runtimeSettings)) {
+    return {
+      tshd: createInsecureClientCredentials(),
+      shared: createInsecureClientCredentials(),
+    };
+  }
+
+  const { certsDir } = runtimeSettings;
+  const [rendererKeyPair, tshdCert, sharedCert] = await Promise.all([
```

Comment:
What happens if any of these promises fail?

Reply:
`createGrpcCredentials` will throw, then after it `getElectronGlobals` will throw. `getElectronGlobals` gets exposed through contextBridge as `window.electron`:

https://github.com/gravitational/webapps/blob/be9639ef963520f678665d8479549b1739f27fc0/packages/teleterm/src/preload.ts#L33

It's used in `boot.tsx`:

https://github.com/gravitational/webapps/blob/be9639ef963520f678665d8479549b1739f27fc0/packages/teleterm/src/ui/boot.tsx#L13-L14

`getElectronGlobals` in `boot.tsx` calls `window.electron` underneath, so it will fail as well, triggering the `catch` branch:

https://github.com/gravitational/webapps/blob/be9639ef963520f678665d8479549b1739f27fc0/packages/teleterm/src/ui/boot.tsx#L28-L30

This will show the error to the user in the actual browser window.